server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+                    logger.info("found initial operator settings file [{}], applying...", settingsFilePath);
+                    // we make a distinction here for startup, so that if we had operator settings before the node started
+                    // we would fail startup.
+                    processFileSettings(settingsFilePath, onStartup);
+                }
+                enableSettingsWatcher(settingsDir);
+            } else {
+                logger.info("operator settings directory [{}] not found, will watch for its creation...", settingsDir);
+                enableSettingsWatcher(environment.configFile());
+            }
+        } catch (Exception e) {
+            if (watchService != null) {
+                try {
+                    this.watchService.close();
+                } catch (Exception ignore) {} finally {
+                    this.watchService = null;
+                }
+            }
+
+            throw new IllegalStateException("unable to launch a new watch service", e);
+        }
+
+        this.watcherThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
+            try {
+                logger.info("file settings service up and running [tid={}]", Thread.currentThread().getId());
+
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    /**
+                     * Reading and interpreting watch service events can vary from platform to platform. E.g:
+                     * MacOS symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_MODIFY:operator
+                     *     ENTRY_CREATE:settings.json
+                     *     ENTRY_MODIFY:settings.json
+                     * Linux in Docker symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_CREATE:operator
+                     * After we get an indication that something has changed, we check the timestamp, file id,
+                     * real path of our desired file.
+                     */
+                    if (Files.exists(settingsDir)) {
+                        try {
+                            Path path = operatorSettingsFile();
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            enableSettingsWatcher(settingsDir);
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("unable to watch or read operator settings file", e);
```

Comment:
Why the "watch" part of this message? There shouldn't even be an exception out of asking to watch the settings dir right?

Reply:
Good point, I'll make it generic, it's a remnant of a refactor I had done.